File: JavascriptAlgorithm/DpPlan/I/deleteAndEarn_740.js
var deleteAndEarn = function (nums) {
  // 2,2,3,3,3,4
  // 0,0,2,3,1
  let max = nums.reduce((max, curr) => Math.max(max, curr), 0);
  let res = 0;
  let all = new Array(max + 1).fill(0);
  nums.map((item) => {
    all[item] = all[item] + 1;
  });
  let dp = [];
  dp[0] = 0;
  dp[1] = all[1];
  all.forEach((item, index) => {
    if (index >= 2) {
      dp[index] = Math.max(dp[index - 1], dp[index - 2] + index * item);
    }
    res = Math.max(res, dp[index]);
  });
  return res;
};

export { deleteAndEarn };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/DpPlan/I/deleteAndEarn_740.java
package algo.tzashinorpu.DpPlan.I;

public class deleteAndEarn_740 {
    public int deleteAndEarn(int[] nums) {
        int max = 0;
        for (int i = 0; i < nums.length; i++) {
            max = Math.max(max, nums[i]);
        }
//        Arrays.stream(nums).max().getAsInt();
        int[] all = new int[max + 1];
        for (int i = 0; i < nums.length; i++) {
            all[nums[i]]++;
        }
        int[] dp = new int[max + 1];
        dp[1] = all[1];
        for (int i = 2; i <= max; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + i * all[i]);
        }
        return dp[max];
    }
}
